src/services/TechRecordsService.ts

```diff
@@ -107,22 +137,138 @@ class TechRecordsService {
-          vrms
+  public updateTechRecord(techRecord: ITechRecordWrapper, msUserDetails: any, files?: string[]) {
+    if (files && files.length) {
+      const promises = [];
+      for (const file of files) {
+        promises.push(this.uploadFile(file, techRecord.vin));
+      }
+      return Promise.all(promises)
+        .then((uploadedData: ManagedUpload.SendData[]) => {
+          const documents: string[] = [];
+          if (uploadedData && uploadedData.length) {
+            for (const uploaded of uploadedData) {
+              const filename = uploaded.Key.split("/")[1];
+              documents.push(filename);
+            }
+          } else {
+            throw new HTTPError(500, HTTPRESPONSE.S3_ERROR);
+          }
+          return this.manageUpdateLogic(techRecord, msUserDetails, documents);
+        })
+        .catch((error: any) => {
+          throw new HTTPError(500, HTTPRESPONSE.S3_ERROR);
         });
-        // Cleaning up unneeded properties
-        delete response.primaryVrm; // No longer needed
-        delete response.secondaryVrms; // No longer needed
-        delete response.partialVin; // No longer needed
-        return response;
+    } else {
+      return this.manageUpdateLogic(techRecord, msUserDetails);
+    }
+  }
+
+  private manageUpdateLogic(techRecord: ITechRecordWrapper, msUserDetails: any, documents?: string[]) {
+    return this.createAndArchiveTechRecord(techRecord, msUserDetails, documents)
+      .then((data: ITechRecordWrapper) => {
+        return this.techRecordsDAO.updateSingle(data)
+          .then((updatedData: any) => {
+            return this.formatTechRecordItemForResponse(updatedData.Attributes);
+          })
+          .catch((error: any) => {
+            throw new HTTPError(error.statusCode, error.message);
+          });
       })
       .catch((error: any) => {
-        throw new HTTPError(error.statusCode, error.message);
+        throw new HTTPError(error.statusCode, error.body);
+      });
+  }
+
+  private createAndArchiveTechRecord(techRecord: ITechRecordWrapper, msUserDetails: any, documents?: string[]) {
+    const isAdrValid = validatePayload(techRecord.techRecord[0]);
+    if (isAdrValid.error) {
+      return Promise.reject({statusCode: 500, body: isAdrValid.error.details});
+    }
+    return this.getTechRecordsList(techRecord.vin, STATUS.ALL)
+      .then((data: ITechRecordWrapper) => {
+        const oldTechRec = this.getTechRecordToArchive(data);
+        oldTechRec.statusCode = STATUS.ARCHIVED;
+        const newRecord: any = _.cloneDeep(oldTechRec);
+        newRecord.statusCode = STATUS.CURRENT;
+        _.merge(newRecord, techRecord.techRecord[0]);
+        if (techRecord.techRecord[0].adrDetails && techRecord.techRecord[0].adrDetails.documents) {
+          newRecord.adrDetails.documents = techRecord.techRecord[0].adrDetails.documents;
+        }
+        if (documents && documents.length) {
+          if (newRecord.adrDetails) {
```

Comment:
No tests for negative cases:
![image](https://user-images.githubusercontent.com/50110006/70146835-ee7a7980-169a-11ea-906a-9532be04dd08.png)

Reply:
fixed